### PR TITLE
Gas refund on purchase

### LINKED ISSUE #7774: Gas refund from lock
We want the locks to be able to refund gas to the `msg.sender` of the transaction.
Right now any address can purchase a membership on behalf of someone else (because the purchase function has a `beneficiary` address. 
We need to add the ability for the lock owner to set a percentage of the key price to be refunded to the sender. By default it should be 0.

This will enable better key purchases through credit cards, as well as cross-chain transactions.

### PR DESCRIPTION
**Pull request by clemsos**
# Description

Set a percentage of the key price to be refunded to the sender on purchase

# Issues

Fixes #7774
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 1]
```diff
@@ -21,6 +20,24 @@ contract MixinPurchase is
 {
   event RenewKeyPurchase(address indexed owner, uint newExpiration);
 
+  event GasRefunded(address indexed receiver, uint refundedAmount, address tokenAddress);
```
Comment (clemsos) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:23]: A specific event is emitted so things can be tracked, but we can also just track transfer event to save gas

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 2]
```diff
@@ -106,6 +123,19 @@ contract MixinPurchase is
...
+    if (_gasRefundPercentage != 0) {
+      uint toRefund = _gasRefundPercentage * pricePaid / 100;
+      if(tokenAddress != address(0)) {
+        IERC20Upgradeable token = IERC20Upgradeable(tokenAddress);
+        token.transferFrom(address(this), msg.sender, toRefund);
+      } else {
+        (bool success, ) = msg.sender.call{value: toRefund}("");
+        require(success, "Refund failed.");
+      }
+      emit GasRefunded(msg.sender, toRefund, tokenAddress);
```
Comment (clemsos) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:137]: Gas refund as a percentage of the key price means that gas is in ETH while key price will be refunded in any sent token. Does that make sense?

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 1 above)
Comment (julien51) [reply smart-contracts/contracts/mixins/MixinPurchase.sol:23]: Let's keep the event for now.

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 2 above)
Comment (julien51) [reply smart-contracts/contracts/mixins/MixinPurchase.sol:137]: Yes and I think that's the only way as we don't know that the lock will have eth to cover the gas... The "submitters" will be able to assess whether it is worth for them to issue the tx based on the conversion rate of a given token.

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 3]
```diff
@@ -21,6 +20,24 @@ contract MixinPurchase is
...
   event RenewKeyPurchase(address indexed owner, uint newExpiration);
 
+  event GasRefunded(address indexed receiver, uint refundedAmount, address tokenAddress);
+
+  uint8 private _gasRefundPercentage = 0; // default to 0
+
+  /**
+  * @dev Set a percentage of the key price to be refunded to the sender on purchase
+  */
+  function setGasRefundPercentage(uint8 _percent) external onlyLockManagerOrBeneficiary {
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:30]:
```suggestion
  function setGasRefundPercentage(uint8 _percent) external onlyLockManager {
```

Let's limit this to the lock manager.

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 4]
```diff
@@ -21,6 +20,24 @@ contract MixinPurchase is
...
+    _gasRefundPercentage = _percent;
+  }
+  
+  /**
+  * @dev Returns percentage be refunded to the sender on purchase
+  */
+  function gasRefundPercentage() external view onlyLockManagerOrBeneficiary returns (uint8 percentage) {
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:37]: I think `uint8` is too small! We need to go up to 10,000 (basis points). Also this should not have the `onlyLockManagerOrBeneficiary` modifier :)

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 5]
```diff
@@ -106,6 +123,19 @@ contract MixinPurchase is
     {
       onKeyPurchaseHook.onKeyPurchase(msg.sender, _recipient, _referrer, _data, inMemoryKeyPrice, pricePaid);
     }
+
+    // refund gas
+    if (_gasRefundPercentage != 0) {
+      uint toRefund = _gasRefundPercentage * pricePaid / 100;
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:129]: `/100` will not work. We need to do this in basis points (check methods like the penalties for an example as it is also a %)

File: smart-contracts/contracts/interfaces/IPublicLock.sol [hunk 6]
```diff
@@ -212,6 +212,16 @@ interface IPublicLock
     bytes calldata _data
   ) external payable;
 
+  /**
+  * @dev Set a percentage of the key price to be refunded to the sender on purchase
+  */
+  function setGasRefundPercentage(uint8 _percent) external;
```
Comment (julien51) [inline smart-contracts/contracts/interfaces/IPublicLock.sol:218]: `uint8` is too small :)

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 3 above)
Comment (julien51) [reply smart-contracts/contracts/mixins/MixinPurchase.sol:30]: Also we should probably check that it is lower (or equal even tho that would not make a lot of sense) to 10,000.

File: smart-contracts/test/Lock/purchaseGasRefund.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    lock = locks.FIRST
+  })
+
+  describe('gas refund percentage', () => {
+    it('get set properly', async () => {
+      await lock.setGasRefundPercentage(25)
+      assert.equal(await lock.gasRefundPercentage(), 25)
+    })
+
+    it('can not be set if caller is not lock manager or beneficiary', async () => {
```
Comment (julien51) [inline smart-contracts/test/Lock/purchaseGasRefund.js:29]: No need to allow the beneficiary...

File: smart-contracts/test/Lock/purchaseGasRefund.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      await lock.addLockManager(accounts[5], { from: accounts[0] })
+      await lock.setGasRefundPercentage(25, { from: accounts[5] })
+      assert.equal(await lock.gasRefundPercentage(), 25)
+    })
+  })
+
+  describe('purchase with gas refund', () => {
+    let tx
+
+    beforeEach(async () => {
```
Comment (julien51) [inline smart-contracts/test/Lock/purchaseGasRefund.js:54]: Let's also write tests for an ERC20 just to make sure :)

**Review by julien51 — CHANGES_REQUESTED** [review]
A few changes are required :)

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 5 above)
Comment (clemsos) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: moved to basis points

File: smart-contracts/test/Lock/purchaseGasRefund.js (hunk 8 above)
Comment (clemsos) [reply smart-contracts/test/Lock/purchaseGasRefund.js]: yep added

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 9]
```diff
@@ -125,8 +127,8 @@ contract MixinPurchase is
     }
 
     // refund gas
-    if (_gasRefundPercentage != 0) {
-      uint toRefund = _gasRefundPercentage * pricePaid / 100;
+    if (_gasRefundBasisPoints != 0) {
+      uint toRefund = _gasRefundBasisPoints * pricePaid / BASIS_POINTS_DEN;
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:131]: 👍 for reusing BASIS_POINTS_DEN!

**Comment by julien51** [comment]
LGTM!

**Review by julien51 — APPROVED** [review]
💯

**Pull request merged**